checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -128,32 +128,80 @@ protected void commonAssignmentCheck(
         // Either a singleton list of the single integer on the right,
         // or a list containing all the values in a constant array.
         List<Integer> rhsValues = new ArrayList<>();
-        // All the values in the must be compile-time constants.
+
+        boolean allValuesConstant = true;
+        // All the values must be compile-time constants to check against the minlen.
         for (ExpressionTree exp : expressions) {
             Integer val = atypeFactory.valMaxFromExpressionTree(exp);
             if (val == null) {
-                super.commonAssignmentCheck(varType, valueExp, errorKey);
-                return;
+                allValuesConstant = false;
             } else {
                 rhsValues.add(val);
             }
         }
 
-        // Actually check that every integer in rhsValues is less than the minlen of each array.
+        boolean conforms = true;
+        // Check whether the expressions are valid for each array listed. Set conforms to true if they aren't.
         for (String arrayName : ltl.getArrays()) {
-            int minLen =
+            // do the samelen check first. Look up all arrays in the SL type and
+            // check if it conforms to one. If so, skip this iteration.
+            List<String> sameLenArrays =
                     atypeFactory
-                            .getMinLenAnnotatedTypeFactory()
-                            .getMinLenFromString(arrayName, valueExp, getCurrentPath());
-            boolean minLenOk = true;
-            for (Integer value : rhsValues) {
-                minLenOk =
-                        minLenOk && ltl.isValuePlusOffsetLessThanMinLen(arrayName, value, minLen);
+                            .getSameLenAnnotatedTypeFactory()
+                            .getSameLensFromString(arrayName, valueExp, getCurrentPath());
+
+            if (sameLenArrays.size() == 0) {
+                if (!allValuesConstant) {
+                    conforms = false;
+                    break;
+                }
             }
-            if (!minLenOk) {
-                super.commonAssignmentCheck(varType, valueExp, errorKey);
-                return;
+
+            // SameLen checks
+            for (ExpressionTree exp : expressions) {
+                AnnotatedTypeMirror currentType = atypeFactory.getAnnotatedType(exp);
```

Comment:
This recomputes the Upper Bound Type type again for each arrayName, but the type doesn't vary.